File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // Prompt for credit card number
    long number = get_long("Number: ");

    // credit card num count
    long temp_n = number;
    int count = 2;
    while (temp_n > 100)
    {
        temp_n = temp_n / 10;
        count++;
    }

    long temp_n2 = number;
    int sum = 0;

    for (int i = 1; i <= count; i++)
    {
        // get the last digit
        int digit = temp_n2 % 10;
        if (i % 2 == 0)
        {
            if (digit * 2 > 9)
            {
                sum += (digit * 2) - 9;
            }
            else
            {
                sum += digit * 2;
            }
        }
        else
        {
            sum += digit;
        }
        temp_n2 /= 10;
    }

    // return if the number is valid
    if (count == 15 && (temp_n == 34 || temp_n == 37) && sum % 10 == 0)
    {
        printf("AMEX\n");
        return 0;
    }
    else if (count == 16 && (temp_n > 50 && temp_n < 56) && sum % 10 == 0)
    {
        printf("MASTERCARD\n");
        return 0;
    }
    else if ((count == 13 || count == 16) && (temp_n / 10 == 4) && sum % 10 == 0)
    {
        printf("VISA\n");
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 1;
    }
}

File: pset6/credit/credit.py
# Prompt for credit card number
while(True):
    number = input("Number: ").strip()
    if number.isnumeric():
        break

# credit card num count
count = len(number)
sum = 0
tmp_count = count - 1
i = 1

while tmp_count >= 0:
    n = number[tmp_count]
    if(i % 2 == 0 ):
        s = int(n) * 2
        
        if s > 9:
            sum += s - 9
        else:
            sum += s
            
    else:
        sum += int(n)
    i += 1
    tmp_count -= 1

valid = sum % 10 == 0

p = number[:2]

# Check if the card is American Express
if count == 15 and (p in ["34","37"]) and valid:
    print("AMEX")
# Check if the card is MasterCard
elif count == 16 and (p in ["51","52","53","54","55"]) and valid:
    print("MASTERCARD")
# Check if the card is Visa
elif (count == 13 or count == 16) and (p[0] == "4") and valid:
    print("VISA")
# Print Invalid if it's not one of the above
else:
    print("invalid")